exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,709 @@
+        2 => Responses {
+            /// Information about transactions, that sent as response to `TransactionsRequest`.
+            TransactionsResponse = 0,
+            /// Information about block, that sent as response to `BlockRequest`.
+            BlockResponse = 1,
+        },
+        /// Exonum node requests.
+        3 => Requests {
+
+            /// Request of some propose which hash is known.
+            ProposeRequest = 0,
+            /// Request of unknown transactions.
+            TransactionsRequest = 1,
+            /// Request of prevotes for some propose.
+            PrevotesRequest = 2,
+            /// Request of peer exchange.
+            PeersRequest = 3,
+            /// Request of some future block.
+            BlockRequest = 4,
+        },
+
+    }
+}
+
+impl Protocol {
+    /// Creates new protocol message.
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn new<T: ProtocolMessage>(
+        message: T,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Protocol {
+        T::into_protocol(Protocol::concrete(message, author, secret_key))
+    }
+
+    /// Creates new protocol message.
+    /// Return concrete `Message<T>`
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn concrete<T: ProtocolMessage>(
+        message: T,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Message<T> {
+        let value = message.serialize().expect("Couldn't serialize data.");
+        let (cls, typ) = T::message_type();
+        let signed = SignedMessage::new(cls, typ, value, author, secret_key);
+        T::into_message_from_parts(message, signed)
+    }
+    /// Checks buffer and return instance of `Protocol`.
+    pub fn verify_buffer(buffer: Vec<u8>) -> Result<Protocol, failure::Error> {
+        let signed = SignedMessage::verify_buffer(buffer)?;
+        Self::deserialize(signed)
+    }
+
+    /// Trying to convert instance of `Protocol` into `Message<RawTransaction>`.
+    pub fn try_into_transaction(self) -> Result<Message<RawTransaction>, failure::Error> {
+        RawTransaction::try_from(self)
+            .map_err(|m| format_err!("Couldn't convert message {:?} into transaction", m))
+    }
+
+    /// Creates new raw transaction message.
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn sign_tx<T>(
+        transaction: T,
+        service_id: u16,
+        public_key: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Message<RawTransaction>
+    where
+        T: Into<TransactionSetPart>,
+    {
+        let set: TransactionSetPart = transaction.into();
+        let raw_tx = RawTransaction::new(service_id, set);
+        Self::concrete(raw_tx, public_key, secret_key)
+    }
+}
+
+impl Requests {
+    /// Returns public key of the message recipient.
+    pub fn to(&self) -> PublicKey {
```

Comment:
In my opinion `receiver` is more clear.

Reply:
I am sure, this question It's out of scope of current design.